### ISSUE TITLE
HttpClient/WinHttpHandler - unexpected delay under heavy load

### ISSUE THREAD

**Issue by jbednar79** [dotnet/corefx#26743]
I've found problem when using the HttpClient class. Under heavy load when I log when the SendAsync() method was called and when it returned the response I can see 3844ms for example. But the party on the other side is reporting 16ms only! By turning on all possible log switches in .NET FW I was able to track to the problematic part - which is in WinHttpHandler, private async void StartRequest(WinHttpRequestState state) method, line with await InternalSendRequestAsync(state); 
As you can see there is a missing .ConfigureAwait(false) which is (probably) doing this unexpected delay under heavy load as it's waiting for original context.

Additional logging follows
Client App [before/after SendAsync method call]:
```
2018-01-31 18:16:51,148; [57]; DEBUG; $$Operation started, Id: 72c574c5c828644b-cba797e5-fcd59352
2018-01-31 18:16:54,992; [111]; TIMING; Id=72c574c5c828644b-cba797e5-fcd59352; Duration=3844; Operation=[POST] [http://<dns/path>]; Message=200 OK 
```

Server App [WebAPI]:
```
2018-01-31 18:16:54,988; [87]; DEBUG; 72c574c5c828644b-fcd59352-0fb4a886	Requested	[POST] <dns/path>	Id=72c574c5c828644b-cba797e5-fcd59352
2018-01-31 18:16:54,988; [87]; DEBUG; 72c574c5c828644b-fcd59352-0fb4a886	Response	Status=200 OK
```

Verbose logging in Client App proves the delay problem is outside our control:
```
    DateTime=2018-01-31T18:16:51.1637939Z
System.Net Verbose: 0 : [3056] HttpWebRequest#33970533::HttpWebRequest(http://<dns/path>#1354891878)
    DateTime=2018-01-31T18:16:51.1637939Z
System.Net Verbose: 0 : [3056] Exiting HttpWebRequest#33970533::HttpWebRequest() 
    DateTime=2018-01-31T18:16:51.1637939Z
System.Net Verbose: 0 : [3056] HttpWebRequest#33970533::HttpWebRequest(uri: 'http://<dns/path>', connectionGroupName: '62473202')
    DateTime=2018-01-31T18:16:51.1637939Z
System.Net Verbose: 0 : [3056] Exiting HttpWebRequest#33970533::HttpWebRequest() 
```

<other log events - see after 3seconds!!>

```
    DateTime=2018-01-31T18:16:54.9765126Z
System.Net Information: 0 : [19808] Associating HttpWebRequest#33970533 with ConnectStream#2301377
    DateTime=2018-01-31T18:16:54.9765126Z
System.Net Information: 0 : [19808] ConnectStream#2301377 - Sending headers
{
id: 72c574c5c828644b-cba797e5-fcd59352
...}
```

[EDIT] Log formatting changes by @karelz

**Comment by davidsh** [comment]
You're correct in that there should be .ConfigureAwait(false) on that line and another one.

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L856

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L863

Feel free to submit a PR if you wish to contribute a fix for this.

**Comment by stephentoub** [comment]
> As you can see there is a missing .ConfigureAwait(false) which is (probably) doing this unexpected delay under heavy load as it's waiting for original context.

This doesn't have anything to do with ConfigureAwait.  Those methods are not returning Task; they're returning an internal awaitable type that already has the ConfigureAwait(false) semantics built into it, meaning it doesn't do anything to track and post back to a captured SynchronizationContext or TaskScheduler.  There's no need for a ConfigureAwait on those, and in fact adding ConfigureAwait to them won't compile as there's no such method for that type.

**Comment by davidsh** [comment]
>they're returning an internal awaitable type that already has the ConfigureAwait(false) semantics built into it, 

Good point.  I forgot about those prior changes.  I wonder if we need some more comments around this code to avoid future confusion as we review code etc.

**Comment by jbednar79** [comment]
I think the discussion is now focused about ConfigureAwait only and problem is unsolved.
I agree I didn't check the method signatures and there is not standard Task but some internal class I don't know. 

But there must be a problem - with lock() inside methods or with tasks or else. Otherwise I don't understand why two subsequent methods calls are after 3000ms delay. Check verbose logging for details - focus on <other log events - see after 3seconds!!> part. Do you have any explanation for this?

**Comment by karelz** [comment]
@jbednar79 it seems you're hitting the problem on .NET Framework (Desktop). Were you able to reproduce it also on .NET Core? Do you have a repro we could look at?
BTW: Do you use localhost by any chance?

**Comment by jbednar79** [comment]
We are using the 'full' .NET FW 4.6.1 as we are hosting Workflow foundation in the same process - this is limiting us from using the .NET Core.
I didn't try localhost as we have our solution hosted on Azure as part of bigger solution. My verbose output was from the machine A1 which connects machines B1 and B2 - hosted on Azure in VMs using the Load Balancer. All applications are Web API 2 hosted on full IIS. 

I don't have any repository with the source code to share. 

We made few optimizations in "servicePointManager" configuration like disabling Nagle algorithm (a lot of small/fast calls], allowing more connections to one DNS, ... because the performance is critical. HttpClient is used properly too (used as singleton).

But it seems this is out of our control. 
If I remember correctly (and based on this conversation) performance counters for the A1 application shows spikes in "Current Queue Length" sometimes (after chart spike long call was reported). But this is something where I cannot say "this is in the HttpClient for sure" - maybe we have some locks in our code.

**Comment by davidsh** [comment]
The traces indicate that this is .NET Framework being used:

>    DateTime=2018-01-31T18:16:54.9765126Z
System.Net Information: 0 : [19808] Associating HttpWebRequest#33970533 with ConnectStream#2301377
    DateTime=2018-01-31T18:16:54.9765126Z
System.Net Information: 0 : [19808] ConnectStream#2301377 - Sending headers
{
id: 72c574c5c828644b-cba797e5-fcd59352
...}

For performance related issues, using Fiddler or turning on System.Net tracing (which uses TraceListener) is not recommended.  Fiddler changes the scenario by adding a proxy server.  And System.Net tracing will impact performance.

Instead, we recommend using Wireshark with a small repro.  It will help illustrate the timings of TCP packets to determine what might be going on.  And it would be best to close all other apps, browsers etc. on the machine to reduce noise in the traces.

**Comment by jbednar79** [comment]
I'll try to capture packets using the Wireshark this week. Thanks!

**Comment by jbednar79** [comment]
So I have data including application log files & system.trace & wireshark capture. Randomly chosen packet which have different times reported by Client / Server.

Same pattern:
**Client App [A1]** - HTTP operation took 1219ms:
```
2018-02-07 16:45:19,500; [37]; TIMING; Id=3ab5f66a2117294e-c474088e-04eef5e9; Duration=1219; Operation=[POST] [http://<dns>]; Message=200 OK
```

**Server App [B1/B2]** - server responded in '0ms' [see time stamps]:
```
2018-02-07 16:45:19,391; [122]; DEBUG; 3ab5f66a2117294e-04eef5e9-0df82479	Requested [POST] http://<dns>
	x-id      :	3ab5f66a2117294e-c474088e-04eef5e9

2018-02-07 16:45:19,391; [122]; DEBUG; 3ab5f66a2117294e-04eef5e9-0df82479	Response	Status=200 OK
```

**Client App [A1] Tracing** - sending headers & data - time gap ~500ms:
```
    DateTime=2018-02-07T16:45:18.9531522Z
System.Net Information: 0 : [7448] HttpWebRequest#62717185 - Request: POST <path> HTTP/1.1
    DateTime=2018-02-07T16:45:18.9531522Z
System.Net Information: 0 : [7448] ConnectStream#41884164 - Sending headers
{
id: 3ab5f66a2117294e-c474088e-04eef5e9
```
after ~500ms
```
#	[Packet #1819]
    DateTime=2018-02-07T16:45:19.3906527Z
System.Net Verbose: 0 : [12864] Exiting HttpWebRequest#62717185::EndGetRequestStream() 	-> ConnectStream#41884164
    DateTime=2018-02-07T16:45:19.3906527Z
System.Net Verbose: 0 : [12864] ConnectStream#41884164::BeginWrite()
    DateTime=2018-02-07T16:45:19.3906527Z
System.Net Verbose: 0 : [12864] Data from ConnectStream#41884164::BeginWrite
    DateTime=2018-02-07T16:45:19.3906527Z
```
**Wireshark capture**
IP 172.20.73 = Client App [A1]; IP 172.20.0.139 = Server App [B1\B2], it's IP of the Azure Load Balancer
Packet **1705** corresponds to headers. Packet **1815** contains data. Packet **1728** ACK **1711**.
Why the packet **1815** is sent after ~500ms?
I can see TCP Retransmission - don't know why now [didn't focus on this but see the direction - it's from the Server to Client - why? And it's not causing the unusual delay.]
![wireshark](https://user-images.githubusercontent.com/18512284/35932910-53d61930-0bf6-11e8-8fd9-65a96b73825f.png)

I can provide full logs - but it contains sensitive data so I don't want to upload them to any public site.

**Comment by jbednar79** [comment]
I've finally found the answer and there is no problem. 
It was caused by default value for min asynchronous I/O threads. When I set this to higher value this pattern is no longer visible. We had around 68 requests per second on 4 core machine.

From the original 19,647 requests was originally around 500 requests with the time between 300 and 7,000ms. After changing the default value I had only around 20 requests with the time between 300 and 689ms (and 15 of them were caused by Garbage Collector run on the server).

Thank you.
